components/autocomplete/index.js

```diff
@@ -330,11 +338,25 @@ export class Autocomplete extends Component {
 		// create a regular expression to filter the options
 		const search = open ? new RegExp( '(?:\\b|\\s|^)' + escapeRegExp( query ), 'i' ) : /./;
 		// filter the options we already have
-		const filteredOptions = open ? filterOptions( search, this.state[ 'options_' + open.idx ] ) : [];
+		let filteredOptions = open ? filterOptions( search, this.state[ 'options_' + open.idx ] ) : [];
 		// check if we should still suppress the popover
 		const suppress = ( open && wasSuppress === open.idx ) ? wasSuppress : undefined;
 		// update the state
 		if ( wasOpen || open ) {
+			if ( open ) {
+				if ( this.props.completers[ open.idx ].setSearch ) {
+					this.props.completers[ open.idx ].setSearch( query, this.props.completers[ open.idx ].getOptions ).then( ( options ) => {
+						const keyedOptions = map( options, ( option, i ) => ( { ...option, key: open.idx + '-' + i } ) );
+						filteredOptions = filterOptions( this.state.search, keyedOptions );
```

Comment:
As this code is asynchronous I would not reuse the `filteredOptions` variable (call it `asyncFilteredOptions` or something) because the code that comes after this async block on the page will not get access to the updated value.

Reply:
Right, good point, I'll rework this part.